tests/jballerina-unit-test/src/test/resources/test-src/query/query-action.bal

```diff
@@ -197,3 +197,61 @@ function testQueryExpressionIteratingOverXMLInFromInQueryAction() returns float
               };
     return total;
 }
+
+function testTypeTestInWhereClause() {
+    int?[] v = [1, 2, (), 3];
+    int total = 0;
+    error? result = from var i in v
+                    where i is int
+                    do {
+                        total += <int>i;
```

Comment:
Any reason you needed the cast?

Reply:
Seems the narrowed env is not used at the do clause

Reply:
`i` is `int?`. `where` doesn't narrow the type?

Reply:
any reason why it shouldn't?

Reply:
There's a spec issue. https://github.com/ballerina-platform/ballerina-spec/issues/703. we don't support this atm?

Reply:
The spec specifies narrowing only with conditional expressions atm. But this will be introduced for `while` statements (https://github.com/ballerina-platform/ballerina-spec/issues/840 - already supported by the implementation) and `where` clauses (https://github.com/ballerina-platform/ballerina-spec/issues/703) for Swan Lake.

I think we can go ahead and add this support now.

Reply:
@lasinicl probably adding the env to the queryEnv stack should solve this. https://github.com/ballerina-platform/ballerina-lang/blob/master/compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/TypeChecker.java#L5177
Please check

Reply:
Will do :)